File: day01/day01.nim
import strutils, sequtils

iterator diffs(values: seq[(int, int)]): bool =
  for (x, y) in values:
    yield x < y

let 
  lines = readFile("day01-test.txt").splitlines()
  depths = lines.map(parseInt)
  zipped = zip(depths[0 .. ^2], depths[1 .. ^1])
var  
  windows = newSeq[int]()
  zippedWindows: seq[(int, int)]

for i in 0 .. depths.len - 3:
  let x = depths[i] + depths[i+1] + depths[i+2]
  windows.add(x)

zippedWindows = zip(windows[0 .. ^2], windows[1 .. ^1])
echo countIt(diffs(zipped), it == true)
echo countIt(diffs(zippedWindows), it == true)
  

File: day01/day01.py
def count_diffs(data: list[int]) -> int:
    """Count all times a next depth measurement is less than the previous one"""
    return sum([x < y for x, y in zip(data, data[1:])])


if __name__ == '__main__':
    with open("day01-test.txt") as f:
        lines = f.read().splitlines()

        # tasks 1
        depths = list(map(int, lines))
        print(count_diffs(depths))

        # task 2
        # Create windows of 3 depth measures
        windows = [sum(depths[i:i+3]) for i in range(len(depths))]
        print(count_diffs(windows))
